### PR TITLE
dkg/bcast: implement reliable broadcast protocol

### PR DESCRIPTION
**Pull request by corverroos**
Implementation of a reliable-broadcast protocol for use in the DKG. 

All peers sign the broadcasted message hash, then the message is broadcasted with those signatures as proof. 

category: feature
ticket: #1889

### CODE REVIEW

File: dkg/bcast/impl.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/obolnetwork/charon/app/errors"
+	"github.com/obolnetwork/charon/app/k1util"
+	"github.com/obolnetwork/charon/p2p"
+)
+
+// New registers a new reliable-broadcast server and returns a reliable-broadcast client function.
+func New(tcpNode host.Host, sender *p2p.Sender, peers []peer.ID, secret *k1.PrivateKey, callback Callback) func(context.Context, proto.Message) error {
```
Comment (corverroos) [inline dkg/bcast/impl.go:21]: I will add tests a bit later today.

File: dkg/bcast/server.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/libp2p/go-libp2p/core/peer"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/obolnetwork/charon/app/errors"
+	pb "github.com/obolnetwork/charon/dkg/dkgpb/v1"
+	"github.com/obolnetwork/charon/p2p"
+)
+
+// NewServer creates a new reliable-broadcast server.
+func NewServer(tcpNode host.Host, singFunc SignFunc, verifyFunc VerifyFunc, callback Callback) *Server {
```
Comment (gsora) [inline dkg/bcast/server.go:18]:
```suggestion
func NewServer(tcpNode host.Host, signFunc SignFunc, verifyFunc VerifyFunc, callback Callback) *Server {
```

File: dkg/bcast/server.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+		p2p.WithDelimitedProtocol(protocolIDMsg),
+	)
+
+	return s
+}
+
+// Server is a reliable-broadcast server.
+type Server struct {
+	callback   Callback
+	singFunc   SignFunc
```
Comment (gsora) [inline dkg/bcast/server.go:43]:
```suggestion
	signFunc   SignFunc
```

**Review by gsora — COMMENTED** [review]
a couple nitpicking comment but overall seems sound, will approve once tests are merged

**Review by gsora — APPROVED** [review]

File: dkg/bcast/impl.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	signFunc := newK1Signer(secret, allow)
+	verifyFunc := newPeerK1Verifier(peers, allow)
+
+	_ = newServer(tcpNode, signFunc, verifyFunc, callback)
+	cl := newClient(tcpNode, peers, p2p.SendReceive, p2p.Send, hashAny, signFunc, verifyFunc)
+
+	return cl.Broadcast
+}
+
+// hashFunc is a function that hashes a any-wrapped protobuf message.
```
Comment (dB2510) [inline dkg/bcast/impl.go:36]:
```suggestion
// hashAny is a function that hashes a any-wrapped protobuf message.
```

File: dkg/bcast/helpers.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+// Callback is a function that is called when a reliably-broadcast message was successfully received.
+type Callback func(ctx context.Context, peerID peer.ID, msgID string, msg proto.Message) error
+
+// signFunc is a function that signs a hash.
+type signFunc func(msgID string, hash []byte) ([]byte, error)
+
+// verifyFunc is a function that verifies a message and its signatures.
+type verifyFunc func(string, *anypb.Any, [][]byte) error
+
+type BroadcastFunc func(ctx context.Context, msgID string, msg proto.Message) error
```
Comment (dB2510) [inline dkg/bcast/helpers.go:31]: godoc please

**Pull request merged**